### ISSUE TITLE
Error in ChannelIndex when slicing AnalogSignal

### ISSUE THREAD

**Issue by JuliaSprenger** [NeuralEnsemble/python-neo#350]
This issue is related to issue #320 and PR #321.
When a ChannelIndex is referring to only a few traces (channels) of the AnalogSignal, slicing the AnalogSignal leads to an IndexError in the ChannelIndex.
```
# Setting up the ChannelIndex and AnalogSignal
#chidx is holding only uneven channel ids
ids = range(10)[::2]
chidx = neo.ChannelIndex(ids, channel_names=['channel {}'.format(i) for i in ids], channel_ids=ids, name='channel test name')
asig = neo.AnalogSignal(np.zeros((100,max(ids)+1))*pq.V, t_start=0*pq.s, t_stop=10*pq.s, sampling_rate=1*pq.Hz)
chidx.analogsignals.append(asig)
chidx.create_relationship()

# actual slicing of the AnalogSignal
asig3 = asig[:,3]
```
leads to 

>   File "python-neo/neo/core/analogsignal.py", line 309, in __getitem__
>     obj.channel_index = self.channel_index.__getitem__(k)
>   File "python-neo/neo/core/channelindex.py", line 159, in __getitem__
>     channel_names = self.channel_names[index]
> IndexError: index 6 is out of bounds for axis 0 with size 5

**Comment by apdavison** [comment]
The problem here is that it is invalid to assign `chidx` to `asig.channel_index`, which is what `create_relationship()` is doing, because the length of `chidx.index` is not the same as `asig.shape[1]`. `create_relationship()` should raise an Exception.

### General context

`AnalogSignal.channel_index` is the _primary_ index for the signal. It is required to have an `index` which is `arange(0, signal.shape[1])`. Its purpose is simply to (optionally) assign ids and labels to each of the channels in the signal.

It is possible to have additional `ChannelIndex` objects that index a subset of the channels in a given `AnalogSignal`, but they should not be assigned to `signal.channel_index`

I think the general solution to this problem is to make all relationship attributes like `AnalogSignal.channel_index` into properties, so that there is no need to call `create_relationship()`, but this is a bit tricky for IOs, so will have to be done very carefully.

**Comment by apdavison** [comment]
Fixed in PR #383
